File: GreenHouseController/greenhousecommon/src/main/java/com/klnvch/greenhousecommon/ui/states/BluetoothStateFragment.java
package com.klnvch.greenhousecommon.ui.states;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.klnvch.greenhousecommon.R;
import com.klnvch.greenhousecommon.models.BluetoothState;
import com.klnvch.greenhousecommon.models.PhoneState;

import java.util.List;

public class BluetoothStateFragment extends ItemStateFragment implements PhoneStateInterface {

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setMessage(null);
    }

    @Override
    public void update(List<PhoneState> states) {
        if (states == null || states.size() == 0) {
            setImage(R.drawable.ic_baseline_bluetooth_disabled_24);
            setDataMissingMessage();
            setAlert();
        } else {
            PhoneState latest = states.get(0);
            int bluetoothState = latest.getBluetoothState();
            String error = latest.getBluetoothError();
            switch (bluetoothState) {
                case BluetoothState.CONNECTED:
                    setImage(R.drawable.ic_baseline_bluetooth_connected_24);
                    setNormal();
                    break;
                case BluetoothState.NOT_AVAILABLE:
                    setImage(R.drawable.ic_baseline_bluetooth_disabled_24);
                    setMessage(R.string.state_message_bluetooth_not_available);
                    setAlert();
                    break;
                case BluetoothState.NOT_ENABLED:
                    setImage(R.drawable.ic_baseline_bluetooth_disabled_24);
                    setMessage(R.string.state_message_bluetooth_not_enabled);
                    setAlert();
                    break;
                case BluetoothState.ADDRESS_NOT_VALID:
                    setImage(R.drawable.ic_baseline_bluetooth_disabled_24);
                    setMessage(error);
                    setAlert();
                    break;
                case BluetoothState.READ_ERROR:
                case BluetoothState.WRITE_ERROR:
                case BluetoothState.SOCKET_ERROR:
                    setImage(R.drawable.ic_baseline_bluetooth_searching_24);
                    setMessage(error);
                    setAlert();
                    break;
                default:
                    setImage(R.drawable.ic_baseline_bluetooth_disabled_24);
                    setAlert();
            }
        }
    }
}

File: GreenHouseController/greenhousecommon/src/main/java/com/klnvch/greenhousecommon/ui/states/BluetoothStateFragment.kt
package com.klnvch.greenhousecommon.ui.states

import android.os.Bundle
import android.view.View
import com.klnvch.greenhousecommon.R
import com.klnvch.greenhousecommon.models.BluetoothState
import com.klnvch.greenhousecommon.models.PhoneState
import com.klnvch.greenhousecommon.ui.states.listeners.BluetoothUptimeInterface
import com.klnvch.greenhousecommon.ui.states.listeners.PhoneStateInterface

class BluetoothStateFragment : ItemStateFragment(),
    PhoneStateInterface, BluetoothUptimeInterface {

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        clearMessage()
    }

    override fun update(states: List<PhoneState>) {
        if (states.isEmpty()) {
            setImage(R.drawable.ic_baseline_bluetooth_disabled_24)
            setDataMissingMessage()
            setAlert()
        } else {
            val latest = states.first()
            val bluetoothState = latest.bluetoothState
            val error = latest.bluetoothError
            when (bluetoothState) {
                BluetoothState.CONNECTED -> {
                    setImage(R.drawable.ic_baseline_bluetooth_connected_24)
                    setMessage(R.string.state_message_bluetooth_connected)
                    setNormal()
                }
                BluetoothState.NOT_AVAILABLE -> {
                    setImage(R.drawable.ic_baseline_bluetooth_disabled_24)
                    setMessage(R.string.state_message_bluetooth_not_available)
                    setAlert()
                }
                BluetoothState.NOT_ENABLED -> {
                    setImage(R.drawable.ic_baseline_bluetooth_disabled_24)
                    setMessage(R.string.state_message_bluetooth_not_enabled)
                    setAlert()
                }
                BluetoothState.ADDRESS_NOT_VALID -> {
                    setImage(R.drawable.ic_baseline_bluetooth_disabled_24)
                    setMessage(error)
                    setAlert()
                }
                BluetoothState.SOCKET_CONNECT_ERROR -> {
                    setImage(R.drawable.ic_baseline_bluetooth_disabled_24)
                    setMessage(R.string.state_message_socket_connect_error)
                    setAlert()
                }
                BluetoothState.READ_ERROR, BluetoothState.WRITE_ERROR, BluetoothState.SOCKET_ERROR -> {
                    setImage(R.drawable.ic_baseline_bluetooth_searching_24)
                    setMessage(error)
                    setAlert()
                }
                else -> {
                    setImage(R.drawable.ic_baseline_bluetooth_disabled_24)
                    setMessage(R.string.state_message_bluetooth_unknown_error)
                    setAlert()
                }
            }
        }
    }

    override fun onBluetoothUptimeUpdated(bluetoothUptime: MutableList<PhoneState>) {
        if (bluetoothUptime.isNotEmpty()) {
            setMessage2("Uptime: " + formatTime(bluetoothUptime.first().time))
        }
    }
}
